### PR TITLE
Add unit tests and fix bugs for sprites

### PR DESCRIPTION
**Pull request by matt439**
# Description

There are currently no unit tests for sprites. A comprehensive set of tests have been added which cover most sprite methods, with the exception of those which rely on a window. In the process of creating the tests, I discovered multiple bugs which have been also been rectified.

### Position and Velocity Type Changes
`_sprite_data` uses the `point_2d` and `velocity_2d` types for `position` and `velocity` respectively. `point_2d` and `velocity_2d` both use the `double` type. `sprite_x`, `sprite_y`, `sprite_set_x`, `sprite_set_y`, `sprite_dx`, `sprite_dy`, `sprite_set_dx`, and `sprite_set_dy` use the `float` type leading to unnecessary casting. Issues arise when the value stored in `point_2d` or `vector_2d`'s double is outside the range of `float`. The methods have been altered to accept and return `double`.

### Sprite-Point Collision Bug at Corners
`sprite_point_collision` calls `sprite_collision_circle`, which then calls `bitmap_cell_circle` which generates a `circle` which does not encompass the sprite. Points tested at the edges of the sprite will incorrectly return no collision. This occurs because the function uses the greater of a sprite's half-width or half-height as the radius. Instead, the length of the hypotenuse of a right triangle formed with the sprite's half-width and half-height as the opposite and adjacent should equal the circle's radius. `bitmap_bounding_circle` has also been corrected as it contains the same bug.

### Point Methods Type Changes
`point_in_circle` casts `double` to `long long` which causes inaccuracy due to the rounding errors. The method has been altered to retain `circle` radius' `double` precision. In addition, `point_in_circle` calls `point_point_distance` which returns `float` instead of `double`. `point_point_distance` has been altered to return `double`, retaining the `double` precision.

### Sprite Speed Methods Type Changes
`sprite_speed` and `sprite_set_speed` return and accept `float`, while calling the methods `vector_magnitude`, `vector_multiply`, and `unit_vector` which all operate at `double` precision. `sprite_speed` and `sprite_set_speed` have been altered to use `double` precision.

### Sprite Rotation Periodicity
`sprite_set_rotation` incorrectly returns `360` when the input value equals `360`. Instead, it should convert `360` to `0` degrees.

### Additional Unit Tests
As the methods in `sprite.cpp` make calls to methods in `images.cpp` and `point_geometry.cpp`, it was necessary to add unit tests for those methods. In the case of images, there is an existing `unit_test_bitmap.cpp` file which I added to. For geometry tests, I created a new `unit_test_geometry.cpp` file. More geometry tests could be added in the future, however at this stage I only focused on those which are relevant to sprites.

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)

## How Has This Been Tested?

Adding the new tests was achieved by running `cmake -G "Unix Makefiles" .` followed by `make` when in the `/splashkit-core/projects/cmake` directory.

## Testing Checklist

- [x] Tested with skunit_tests

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

### CODE REVIEW

**Review by breezy-codes — APPROVED** [review]
I’ve completed my review of this PR and have the following feedback:

### Sprite Tests:
- The sprite tests work well, and the collision detection is implemented effectively. 
- The circle and rectangle outlines around the sprite were very helpful for verifying and visualising the collision functionality. 
- The line within the sprite was a nice touch, it made it easier to evaluate how speed and velocity are handled.

### Geometry Tests:
- The geometry tests are well-structured. I appreciate that the windows don’t all open simultaneously; it’s much easier to review each window individually before moving to the next.
- While I’m not deeply familiar with how the geometry functions work within SplashKit, I noticed something in the triangle test. Is the circle on each triangle supposed to move simultaneously? As I moved my mouse, multiple circles appeared to track it. If this behaviour is intended, feel free to disregard this comment.

### Unit Tests:
- The unit tests are thorough and extensive, covering a wide range of functionality and edge cases. Excellent work here!

### Other Notes:
- Great attention to detail with the float-to-sprite changes. I’ve reviewed these updates and confirmed that the `.cpp` and `.h` files align correctly.
- I also noticed how the changes flowed into related functions, like updating `point_point_distance` in `point_in_circle`. The use of `sqrt` is a solid choice, this has inspired me to tweak a similar part of my own PR.


Note - All tests were done on Manjaro Arch Linux.

**Review by NoahJCross — APPROVED** [review]
I've reviewed this PR and approve of these changes.  

### Code Quality & Fixes
- Excellent work converting float to double types for better precision. I definitely prefer these more precise calculations, especially for distance and geometric computations.
- Important bug fix for sprite-point collision detection at corners using proper hypotenuse calculation
- Well-thought-out type consistency improvements across related methods

### Testing
- Comprehensive new unit test coverage for sprites
- Tests appear to cover both normal cases and edge cases effectively

### Documentation
- Clear explanation of all changes and their rationale
- Well-documented bug fixes with technical details
- Good breakdown of type conversion changes and their impact

### Personal Experience
I've encountered these exact type inconsistency issues while implementing the rust bindings. Specifically:
- mouse_x() returns f32 but is used in update_sprite to create a circle_at which expects f64 parameters
- vector_from_angle takes two f64 parameters but passes them to sine and cosine functions which expect f32

This does work since the FFI bindings are passing these values to the c code which automatically casts them to f32 but if a user for example wanted to draw a shape at their cursor position using mouse_x and mouse_y, languages that dont automatically cast to 32 byte floating numbers would need to explicitly cast the return of mouse_x and mouse_y to 64 bytes before passing it to the drawing function.

The changes improve both accuracy and consistency while adding much-needed test coverage. The breaking changes are necessary and well-justified for improving the overall reliability of the sprite system.

File: coresdk/src/coresdk/point_geometry.cpp [hunk 1]
```diff
@@ -138,7 +138,12 @@ namespace splashkit_lib
 
     bool point_in_circle(const point_2d &pt, const circle &c)
     {
-        return point_point_distance(c.center, pt) <= abs((long long)c.radius);
+        double radius = c.radius;
+        if (c.radius < 0.0)
+        {
+            radius = -c.radius;
+        }
+        return point_point_distance(c.center, pt) <= radius;
```
Comment (WhyPenguins) [inline coresdk/src/coresdk/point_geometry.cpp:146]: Change looks good, but I believe this could be more simply written as `return point_point_distance(c.center, pt) <= fabs(c.radius);`.

File: coresdk/src/coresdk/sprites.cpp [hunk 2]
```diff
@@ -1192,18 +1192,29 @@ namespace splashkit_lib
...
+        }
+        if (value == 0.0)
+        {
+            s->velocity = vector_to(0.0, 0.0);
+        }
+        else if (vector_magnitude(s->velocity) == 0.0)
+        {
+            s->velocity = vector_to(1.0, 0.0);
+        }
+        s->velocity = vector_multiply(unit_vector(s->velocity), value);
```
Comment (WhyPenguins) [inline coresdk/src/coresdk/sprites.cpp:1217]: I'm unsure about this one. The explicit `if (value == 0.0)` doesn't have any effect, and I'm not sure forcing movement to the right(?) if the sprite doesn't have a direction yet is intuitive. 

The original behaviour of having no effect if the sprite doesn't already have a velocity isn't necessarily ideal, but not sure moving in an arbitrary direction is great either. 

Maybe adding a line to the documentation explaining what the function does if the sprite isn't already moving would be better?

**Review by WhyPenguins — COMMENTED** [review]
Nice work on this! It looks great for the most part, just had a couple of questions/thoughts on some bits :smiley:

File: coresdk/src/coresdk/sprites.cpp (hunk 2 above)
Comment (matt439) [reply coresdk/src/coresdk/sprites.cpp]: This was a workaround for the behavior when calling `sprite_set_speed` when velocity is a zero vector. I have altered the unit test to set a velocity instead of speed. This allows for a much simpler `sprite_set_speed`.

File: coresdk/src/coresdk/point_geometry.cpp (hunk 1 above)
Comment (matt439) [reply coresdk/src/coresdk/point_geometry.cpp]: This has been fixed.

**Comment by matt439** [comment]
Thanks for the feedback. I have made adjustments to fix the issues.
